src/driver-openinput.c

```diff
@@ -0,0 +1,483 @@
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "config.h"
+
+#include "libratbag-private.h"
+
+
+/* reports */
+#define OI_REPORT_SHORT		0x20
+#define OI_REPORT_LONG		0x21
+#define OI_REPORT_SHORT_SIZE	8
+#define OI_REPORT_LONG_SIZE	32
+
+#define OI_REPORT_MAX_SIZE		OI_REPORT_LONG_SIZE
+#define OI_REPORT_DATA_INDEX		3
+#define OI_REPORT_DATA_MAX_SIZE		OI_REPORT_LONG_SIZE - OI_REPORT_DATA_INDEX
+
+/* protocol function pages */
+#define OI_PAGE_INFO			0x00
+#define OI_PAGE_GIMMICKS		0xFD
+#define OI_PAGE_DEBUG			0xFE
+#define OI_PAGE_ERROR			0xFF
+
+/* info page (0x00) functions */
+#define OI_FUNCTION_VERSION			0x00
+#define OI_FUNCTION_FW_INFO			0x01
+#define OI_FUNCTION_SUPPORTED_FUNCTION_PAGES	0x02
+#define OI_FUNCTION_SUPPORTED_FUNCTIONS		0x03
+
+/* error page (0xFF) */
+#define OI_ERROR_INVALID_VALUE		0x01
+#define OI_ERROR_UNSUPPORTED_FUNCTION	0x02
+#define OI_ERROR_CUSTOM			0xFE
+
+
+static unsigned int report_rates[] = { 125, 250, 500, 750, 1000 };
+
+struct openinput_drv_data {
+	unsigned int num_profiles;
+	unsigned int num_resolutions;
+	unsigned int num_buttons;
+	unsigned int num_leds;
+	unsigned int fw_major;
+	unsigned int fw_minor;
+	unsigned int fw_patch;
+	uint64_t supported;
+};
+
+struct oi_report_t {
+	uint8_t id;
+	uint8_t function_page;
+	uint8_t function;
+	uint8_t data[29];
+} __attribute__((__packed__));
+
+
+/* gcc is being picky and asking for the prototypes to bve defined (?) */
+const char* openinput_function_page_get_name(uint8_t page);
+const char* openinput_function_get_name(uint8_t page, uint8_t function);
+
+#define CASE_RETURN_STRING(a) case a: return #a; break
+const char* openinput_function_page_get_name(uint8_t page)
+{
+	static char numeric[16];
+	char *str;
+
+	switch(page)
+	{
+	CASE_RETURN_STRING(OI_PAGE_INFO);
+	CASE_RETURN_STRING(OI_PAGE_GIMMICKS);
+	CASE_RETURN_STRING(OI_PAGE_DEBUG);
+	CASE_RETURN_STRING(OI_PAGE_ERROR);
+	default:
+		sprintf_safe(numeric, "0x%02x", page);
+		str = numeric;
+		break;
+	}
+
+	return str;
+}
+
+const char* openinput_function_get_name(uint8_t page, uint8_t function)
```

Comment:
nitpick on style: at least the first file I looked at (`driver-hidpp10.c`) uses:
```
const char*
function name(...)
```